scratch/controller/controller.cpp

```diff
@@ -0,0 +1,127 @@
+    int16_t left, right;
+};
+
+//Joystick values range from -32767 to +32767
+//Used formula scaled_val = (new_max - new_min) / (old_max - old_min) * (v - old_min) + new_min
+//Can the rover handle the calibration side instead?
+int16_t recalibrate(int16_t axis_value){
+    int16_t joy_stick_abs = axis_value >=0 ? axis_value : axis_value*-1;
+    int16_t new_axis_value = axis_value;
+    float percentage = (float(MAX_POS)-0)/(MAX_POS-MIN_POS);
+    new_axis_value = percentage*(joy_stick_abs-MIN_POS);
+    if (axis_value < 0){
+        return -new_axis_value;
+     }
+        return new_axis_value;
+};
+
+int main(int argc, char *argv[]) {
+    network::Connection conn;
+    network::Error error;
+    if ((error = network::connect(&conn, "192.168.1.1", 45545, 45545)) != network::Error::OK) {
+        fprintf(stderr, "[!] Failed to connect to rover!\n");
+          return 1;
+    }
+    network::MovementMessage currentMovement = {0};
+    network::Buffer* outgoing;
+    //Device input defaults to js0
+    const char* device_name = "/dev/input/js0";
+    if (argc == 2) 
+      device_name = argv[1];
+    //Controller file descriptor. O_NONBLOCK mdoe ensures multiple events can
+    //be read and added to diver queue
+    int fd = open(device_name, O_RDONLY | O_NONBLOCK); 
+    if (fd == -1){ 
+      fprintf(stderr, "[!] Failed to find controller!\n");
+      return 1;  
+    }
+    struct js_event event;
+    //Main loop
+    while (true) {
+        //Loop until event queue is empty, resulting in -1
+        while (read(fd, &event, sizeof(event)) > 0) { 
+            unsigned short button_number;
+            const char* button_state;
+            switch (event.type){ 
+                //Unused button event handler
+  	            case JS_EVENT_BUTTON:
+                   button_number = event.number;
+                   //Pressed is 1, released is 0
+                   button_state =  event.value ? "pressed" : "released"; 
+
+                    break;
+                //Only care about the left and right Y values for axis events
+                case JS_EVENT_AXIS:
+                    switch (event.number) {
+                        case JOYSTICK_LEFT_Y:
+                            //event.value is defined as a signed 16bit value
+                            currentMovement.left = event.value;
+                            outgoing = network::get_outgoing_buffer();
+                            network::serialize(outgoing, &currentMovement); 
+                            network::queue_outgoing(&conn, network::MessageType::MOVEMENT, outgoing);
+                            break;
+                        case JOYSTICK_RIGHT_Y:
+                            currentMovement.right = event.value;
+                            outgoing = network::get_outgoing_buffer();
+                            network::serialize(outgoing, &currentMovement); 
+                            network::queue_outgoing(&conn, network::MessageType::MOVEMENT, outgoing);
```

Comment:
I like what you are doing with the `currentMovement` variable to store the movement state. Currently, you are sending a movement message every time the controller values change. That is great, however I would like to send controller updates at a fixed rate (say, every 100-200ms, so between 5 and 10 a second should be fine). So you'll need to keep track of the time since you last sent an update, and only send a message if that time is greater or equal to the wait time. Use the `<chrono>` library for timing (http://www.cplusplus.com/reference/chrono/).